docs/assertions_guide.md

```diff
@@ -0,0 +1,315 @@
+  assert(0 < mat->height && mat->height <= MAX_MATRIX_HEIGHT);
+}
+```
+
+Then, call it at as needed to proactively sanity check:
+
+```c++
+// REQUIRES: mat points to a valid Matrix
+// EFFECTS:  Returns a pointer to the last element (at the
+//           highest possible row/column) in the Matrix.
+int * Matrix_last(Matrix* mat) {
+  Matrix_check_invariants(mat);
+  return &mat->data[mat->height * mat->width - 1];
+}
+```
+{: data-highlight="5" }
+
+In the example above, you could think of this as a check on the REQUIRES
+clause that specifies `mat` must point to a _valid_ `Matrix`. This keeps
+us from wasting time if, for example, `Matrix_last()` appears to be broken
+but the problem was really that we forgot to call `Matrix_init()` previously.
+
+#### Example: `List` ADT
+{: .primer-spec-toc-ignore }
+
+In a C++ style class, implement `check_invariants()` as a private member function. Call it after initializing member variables in the constructor and any time you make changes to your data representation (i.e. the member variables).
+
+Here's an example for a complex data structure called a linked list, in which individual nodes store data and connect to their neighbors via pointers:
+
+```c++
+class List {
+private:
+  // Lists contain nodes, connected in a chain using pointers
+  struct Node {
+    Node *next;
+    Node *prev;
+    int datum;
+  };
+
+  // Member variables
+  Node *first;
+  Node *last;
+  int num_nodes;
+
+  // Check represenation invariants with assert
+  void check_invariants() {
+    // Either it's empty with null first/last pointers,
+    // or it's not empty and neither should be null
+    assert(!first && !last || first && last);
+
+    // There's nothing before the first node
+    // and nothing after the last node.
+    assert(first->prev == nullptr);
+    assert(last->next == nullptr);
+
+    // Manually counting the number of nodes matches
+    // the number we're tracking in num_nodes
+    int count = 0;
+    for(Node *n = first; n; n = n->next) {
+      ++count;
+    }
+    assert(count == num_nodes);
+  }
+
+public:
+
+  // Default constructor. Creates an empty list.
+  List : first(nullptr), last(nullptr), num_nodes(0) {
+    check_invariants(); // ensure our list is set up
+  }
+
+  //EFFECTS:  inserts datum into the front of the list
+  void push_front(const T &datum) {
+
+    // A bunch of complicated code that rearranges
+    // pointers to add a node to the list
+    // ...
+
+    check_invariants(); // make sure we didn't screw up the pointers
+  }
+};
+```
+{: data-highlight="15-33,39,49" }
+
+## Debug Failed Assertions
+
+Let's say you're running `Image_public_test.exe` on project 2 and
+one of your `Matrix_at()` assertions fails:
+
+```console
+$ ./Image_public_tests
+Running test: test_image_basic
+Image_public_test.exe: Matrix.cpp:72: int* Matrix_at(Matrix*, int, int): Assertion `0 <= row && row < Matrix_height(mat)' failed.
+Aborted
+```
+{: data-highlight="16,39,49" data-variant="legacy" }
```

Comment:
```suggestion
{: data-highlight="16,39,49" }
```

Enhanced code blocks are pretty nice!  Consider removing the `legacy` option.

Reply:
I had been using this because I didn't want line numbers, but I discovered that there is now a specific "no-line-numbers" variant! :D